lntemp/node/watcher.go

```diff
@@ -0,0 +1,679 @@
+
+	// watchCloseChannel specifies that this is a request to watch a close
+	// channel event.
+	watchCloseChannel
+
+	// watchPolicyUpdate specifies that this is a request to watch a policy
+	// update event.
+	watchPolicyUpdate
+
+	// TODO(yy): remove once temp tests is finished.
+	DefaultTimeout = lntest.DefaultTimeout
+)
+
+// closeChanWatchRequest is a request to the lightningNetworkWatcher to be
+// notified once it's detected within the test Lightning Network, that a
+// channel has either been added or closed.
+type chanWatchRequest struct {
+	chanPoint wire.OutPoint
+
+	chanWatchType chanWatchType
+
+	eventChan chan struct{}
+
+	advertisingNode    string
+	policy             *lnrpc.RoutingPolicy
+	includeUnannounced bool
+}
+
+// nodeWatcher is a topology watcher for a HarnessNode. It keeps track of all
+// the topology updates seen in a given node, including NodeUpdate,
+// ChannelEdgeUpdate, and ClosedChannelUpdate.
+type nodeWatcher struct {
+	// rpc is the RPC clients used for the current node.
+	rpc *rpc.HarnessRPC
+
+	// state is the node's current state.
+	state *State
+
+	// chanWatchRequests receives a request for watching a particular event
+	// for a given channel.
+	chanWatchRequests chan *chanWatchRequest
+
+	// For each outpoint, we'll track an integer which denotes the number
+	// of edges seen for that channel within the network. When this number
+	// reaches 2, then it means that both edge advertisements has
+	// propagated through the network.
+	// openChanWatchers  map[wire.OutPoint][]chan struct{}
+	openChanWatchers *sync.Map
+
+	// closeChanWatchers map[wire.OutPoint][]chan struct{}
+	closeChanWatchers *sync.Map
+
+	wg sync.WaitGroup
+}
+
+func newNodeWatcher(rpc *rpc.HarnessRPC, state *State) *nodeWatcher {
+	return &nodeWatcher{
+		rpc:               rpc,
+		state:             state,
+		chanWatchRequests: make(chan *chanWatchRequest, 100),
+		openChanWatchers:  &sync.Map{},
+		closeChanWatchers: &sync.Map{},
+	}
+}
+
+// GetNumChannelUpdates reads the num of channel updates inside a lock and
+// returns the value.
+func (nw *nodeWatcher) GetNumChannelUpdates(op wire.OutPoint) int {
+	result, ok := nw.state.numChanUpdates.Load(op)
+	if ok {
+		return result.(int)
+	}
+	return 0
+}
+
+// GetPolicyUpdates returns the node's policyUpdates state.
+func (nw *nodeWatcher) GetPolicyUpdates(op wire.OutPoint) PolicyUpdate {
+	result, ok := nw.state.policyUpdates.Load(op)
+	if ok {
+		return result.(PolicyUpdate)
+	}
+	return nil
+}
+
+// GetNodeUpdates reads the node updates inside a lock and returns the value.
+func (nw *nodeWatcher) GetNodeUpdates(pubkey string) []*lnrpc.NodeUpdate {
+	result, ok := nw.state.nodeUpdates.Load(pubkey)
+	if ok {
+		return result.([]*lnrpc.NodeUpdate)
+	}
+	return nil
+}
+
+// WaitForNumChannelUpdates will block until a given number of updates has been
+// seen in the node's network topology.
+func (nw *nodeWatcher) WaitForNumChannelUpdates(op wire.OutPoint,
```

Comment:
I think there's another generalization we can extract here. 

```go
func (nw *nodeWatcher) waitForNodeState(statePredicate func(*nodeWatcher) bool) {
	ticker := time.NewTicker(wait.PollInterval)
	timer := time.After(DefaultTimeout)
	defer ticker.Stop()

	for {
		fin := statePredicate(nw)

		select {
		// Check the num every 200ms.
		case <-ticker.C:
			if fin {
				return nil
			}

		case <-timer:
			return fmt.Errorf("timeout waiting for num channel "+
				"updates, want %d, got %d", expected, num)
		}
	}
}
```

Then we can expose public funcs like `WaitForNumChannelUpdates` to give more context to the user where in this case `statePredicate` is:
```go
func numEnoughNumUpdates(nw *nodeWatcher, expected int) bool {
    num := nw.GetNumChannelUpdates(op)

    return num >= expect
}
```